pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,661 @@
+		return
+	}
+
+	if workspaceName == tenancyv1alpha1.RootCluster &&
+		requestInfo.IsResourceRequest &&
+		requestInfo.Verb == "get" &&
+		requestInfo.APIGroup == tenancyv1beta1.SchemeGroupVersion.Group &&
+		requestInfo.Resource == "workspaces" &&
+		requestInfo.Name == "~" {
+		var homeWorkspace *tenancyv1beta1.Workspace
+		homeWorkspaceFullName := h.getHomeWorkspaceName(user.GetName())
+		homeClusterWorkspace, err := h.getLocalClusterWorkspace(homeWorkspaceFullName)
+		if err != nil {
+			responsewriters.InternalError(rw, req, err)
+			return
+		}
+		if homeClusterWorkspace != nil {
+			// The home workspace for the user exists, so let's return it to the user if he request to see its definition
+
+			homeWorkspace = &tenancyv1beta1.Workspace{}
+			projection.ProjectClusterWorkspaceToWorkspace(homeClusterWorkspace, homeWorkspace)
+		} else {
+			// Test if the user has the right to see his Home workspace even though it doesn't exists
+			// => test the get verb on the clusterworkspaces/workspace subresource named ~ in the root workspace.
+
+			attributes := homeWorkspaceAuthorizerAttributes(user, "get")
+			decision, reason, err := h.authz.Authorize(ctx, attributes)
+			if err != nil {
+				responsewriters.InternalError(rw, req, err)
+				return
+			}
+			if decision != authorizer.DecisionAllow {
+				responsewriters.Forbidden(ctx, attributes, rw, req, reason, homeWorkspaceCodecs)
+				return
+			}
+
+			parent, name := homeWorkspaceFullName.Split()
+			homeWorkspaceURL := &url.URL{Scheme: "https", Host: req.Host, Path: homeWorkspaceFullName.Path()}
+			homeWorkspace = &tenancyv1beta1.Workspace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        name,
+					ClusterName: parent.String(),
+				},
+				Spec: tenancyv1beta1.WorkspaceSpec{
+					Type: tenancyv1alpha1.ClusterWorkspaceTypeReference{
+						Name: tenancyv1alpha1.ClusterWorkspaceTypeName("Home"),
+						Path: tenancyv1alpha1.RootCluster.String(),
+					},
+				},
+				Status: tenancyv1beta1.WorkspaceStatus{
+					URL: homeWorkspaceURL.String(),
+				},
+			}
+		}
+
+		responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
+		return
+	}
+
+	needsAutomaticCreation, workspaceType := h.needsAutomaticCreation(workspaceName)
+	if needsAutomaticCreation {
+		isHome := workspaceType == "Home"
+		var retryAfterSeconds int
+		var retryLater bool
+		var localCheckError error
+		var foundLocally bool
+		foundLocally, retryLater, retryAfterSeconds, localCheckError = h.checkWorkspaceLocally(workspaceName, isHome, user.GetName())
+		if localCheckError != nil {
+			responsewriters.InternalError(rw, req, localCheckError)
+			return
+		}
+
+		if !foundLocally {
+			// Home or bucket workspace not found in the local informer
+			// Let's try to create it
+
+			// But first check we have the right to do so.
+			attributes := homeWorkspaceAuthorizerAttributes(user, "create")
```

Comment:
move to the Authorize call

Reply:
We also use it when returning the `Forbidden` error in case of userName mismatch. This is why it is created a bit earlier.